xds/internal/xdsclient/clientimpl_watchers.go

```diff
@@ -103,3 +107,80 @@ func (c *clientImpl) WatchEndpoints(clusterName string, cb func(xdsresource.Endp
 		unref()
 	}
 }
+
+// WatchResource uses xDS to discover the resource associated with the provided
+// resource name. The resource type implementation determines how xDS requests
+// are sent out and how responses are deserialized and validated. Upon receipt
+// of a response from the management server, an appropriate callback on the
+// watcher is invoked.
+func (c *clientImpl) WatchResource(rType xdsresource.Type, resourceName string, watcher xdsresource.GenericResourceWatcher) (cancel func()) {
+	// Return early if the client is already closed.
+	//
+	// The client returned from the top-level API is a ref-counted client which
+	// contains a pointer to `clientImpl`. When all references are released, the
+	// ref-counted client sets its pointer to `nil`. And if any watch APIs are
+	// made on such a closed client, we will get here with a `nil` receiver.
+	if c == nil || c.done.HasFired() {
+		c.logger.Warningf("Watch registered for name %q of type %q, but client is closed", rType.TypeEnum().String(), resourceName)
+		return func() {}
+	}
+
+	if err := c.resourceTypes.maybeRegister(rType); err != nil {
+		c.serializer.Schedule(func(context.Context) { watcher.OnError(err) })
+		return func() {}
+	}
+
+	return func() {} // TODO: replace this with the code below.
+	/*
```

Comment:
Generally: don't check in commented-out code.  Save it elsewhere if you must. Also, please make it clear when we should make the change.

Reply:
I left in in there specifically to give you an idea of what should eventually happen. I should have written a comment for that instead. Done that now.